documentation/design_proposal/002_grafana_v1beta1.md

```diff
@@ -0,0 +1,191 @@
+}
+```
+
+## Proposal
+
+To get rid of required fields we could use a similar strategy as [Banzai Cloud](https://github.com/banzaicloud/operator-tools/blob/2189d8efc3856efd4a7c7fbb28b7cba9a977d0bd/pkg/typeoverride/override.go) does.
+
+I know much of these things is already done in our repo.
+There have been problems with merging specific volumes and such inside the deployment and giving the option to override makes this complex.
+We might just be happy with the current solution that we have and just make sure that all resources part of the object is defined in our code.
+
+```.go
+type GrafanaSpec struct {
+ Config                     GrafanaConfig            `json:"config"`
+ DashboardLabelSelector     []*metav1.LabelSelector  `json:"dashboardLabelSelector,omitempty"`
+ Ingress                    *netv1.Ingress           `json:"ingress,omitempty"`
+ RouteOpenShift             *v12.Route               `json:"routeOpenShift,omitempty"`
+ Service                    *v1.Service              `json:"service,omitempty"`
+ Deployment                 *typeoverride.Deployment `json:"deployment,omitempty"`
+ ServiceAccount             *v1.ServiceAccount       `json:"serviceAccount,omitempty"`
+ Client                     *GrafanaClient           `json:"client,omitempty"`
+ DashboardNamespaceSelector *metav1.LabelSelector    `json:"dashboardNamespaceSelector,omitempty"`
+ Jsonnet                    *JsonnetConfig           `json:"jsonnet,omitempty"`
+ GrafanaContainer           *GrafanaContainer        `json:"grafanaContainer,omitempty"`
+}
+
+// GrafanaContainer provides a means to configure the grafana container
+type GrafanaContainer struct {
+ BaseImage         string                   `json:"baseImage,omitempty"`
+ Resources         *v1.ResourceRequirements `json:"resources,omitempty"`
+ ReadinessProbe    *v1.Probe                `json:"readinessProbe,omitempty"`
+ LivenessProbeSpec *v1.Probe                `json:"livenessProbe,omitempty"`
+}
+```
+
+Instead of setting `grafana.spec.baseImage` you would have to do `grafana.spec.grafanaContainer.baseImage`.
+
+Checkout the [grafana_types](002_grafana_types.go) file for a better example. Just overwrite `api/integreatly/v1alpha1/grafana_types.go` with this file.
+
+### Change in defaults
+
+#### Default Security Context
+
+Today we have a number of defaults that would be good to apply to the operator it self and the grafana deployment
+to make it as secure as possible out of the box. One of the problems with this in openshift [SCC](https://docs.openshift.com/container-platform/4.9/authentication/managing-security-context-constraints.html)
+and we have to add some check to see if we are deploying in a openshift environment or not.
+
+One of the configs we should asspire to is something like this in our container seucirty context.
+Openshift don't support providing `runAsGroup` and `runAsUser`, `runAsNonRoot`might work since it gets user and group from openshift by default.
```

Comment:
I think kubebuilder also allows us to set default values from the CSV level, which might also be beneficial when dealing with those fields i.e automatically populating suggested defaults, without having to programme that in the operator itself, which would be pretty neat

Reply:
That would be really nice. I have been looking around to see if I can find anything about it but I'm unable to do so.
It would be nice to add to the document, at the same time it's a implementation detail and not supper important from a user point of view. But still it would be good to know if it's possible.